fog/block_provider/src/lib.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) 2018-2023 The MobileCoin Foundation
+
+//! Abstractions for getting ledger db data, either from a local LedgerDB or a
+//! remote mobilecoind. Geared towards the specific data fog services require.
+
+mod error;
+mod local;
+mod mobilecoind;
+
+use dyn_clone::DynClone;
+use mc_blockchain_types::{Block, BlockData, BlockIndex};
+use mc_crypto_keys::CompressedRistrettoPublic;
+use mc_fog_api::ledger::TxOutResult;
+use mc_transaction_core::tx::{TxOut, TxOutMembershipProof};
+use mc_watcher_api::TimestampResultCode;
+use std::time::Duration;
+
+pub use error::Error;
+pub use local::LocalBlockProvider;
+pub use mobilecoind::MobilecoindBlockProvider;
+
+pub trait BlockProvider: DynClone + Send + Sync {
```

Comment:
small question: what about re-using the Ledger trait from `mc-ledger-db` crate?
that is how i imagined a change like this would go

Reply:
There are a few reasons why I ended not doing that:
1. The main reason is that the ops team has preference to have the option to load-balance against multiple `mobilecoind`s for redundancy. This means that you might hit different `mobilecoind` instances that are not synced to the exact same block when doing a few consecutive calls (e.g. `num_blocks` + `num_txos` + `get_block` + `get_block_contents`) which could result in weird behavior. A potential way around it would be to add the methods in the `BlockProvider` trait to the `Ledger` trait, but since the assortment of data that fog requires is pretty specific, I am not sure its worth making that `Trait` even more elaborate than it already is.
2. The watcher stuff - Fog cares about timestamps and none of that is represented in the `Ledger` trait. None of the watcher stuff currently has a trait, so it will require adding that.
3. Pretty minor but the `Ledger` trait has a lot of methods that Fog doesn't need, so the `mobilecoind` implementation would either need to return an error on panic on those, but without proper unit test coverage for fog using the mobilecoind backend, if we ever make any changes and start using a `Ledger` API we have not used before/implemented, we'll only find out during CD/live network testing

I don't think the `BlockProvider` is a particularly elegant solution, but when weighing it against these reasons I felt like it's slightly less prone to misbehavior. I'm open for feedback or course correction if you think that makes sense to do. Thanks for looking at this!